\documentclass[a4paper]{article}
%\VignetteIndexEntry{An introduction to WhopGenome}
\usepackage{Sweave}
\usepackage{color}
\usepackage{verbatim}
\usepackage[utf8]{inputenc}
\title{Using WhopGenome}
\author{Ulrich WittelsbÃ¼rger}
\newenvironment{rcode}
{%
\color{blue}
\verbatim
}
{% This is the end code
\endverbatim
}

\begin{document}


%
%
%
\maketitle

%
%	using VCF files
%
\section{Handling VCF files}

\subsection{Opening a VCF file}
Suppose we want to have a look at some of the 1000 Genome Project data for chromosome Y,
available on the EBI ftp at ftp:\/\/ftp.1000genomes.ebi.ac.uk\/vol1\/ftp\/phase1\/analysis\_results\/integrated\_call\_sets\/ALL.chrY.phase1\_samtools\_si.20101123.snps.low\_coverage.genotypes.vcf.gz .

\begin{rcode}
> library(WhopGenome) # load WhopGenome first
> 
> # Make things a bit easier by having a variable instead of a lengthy filename
> yvcffile <- "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/analysis_results/integrated_call_sets/ALL.chrY.phase1_samtools_si.20101123.snps.low_coverage.genotypes.vcf.gz"
> 
> yvcfh <- vcf_open(yvcffile)
[get_local_version] downloading the index file...
(II) VCF version is 4.1
vcff::open : file opened, contains 526 samples
> 
\end{rcode}


As a demonstration for vcf\_reopen, let's quit R and restart it:

\begin{rcode}
> q()
Save workspace image? [y/n/c]: y
user@computer:~\$ R

R version 3.0.1 (2013-05-16) -- "Good Sport"
Copyright (C) 2013 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

[...]
[Previously saved workspace restored.]

> library(WhopGenome)
> yvcfh
<pointer: (nil)>
attr(,"VCF.filename")
[1] "ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/analysis_results/integrated_call_sets/ALL.chrY.phase1_samtools_si.20101123.snps.low_coverage.genotypes.vcf.gz"
>
\end{rcode}

As you can see, the filehandle variable \textit{yvcfh} is present but the data it refers to isn't anymore, because it was managed by WhopGenome and not R. This is the typical
use-case for vcf\_reopen (and tabix\_reopen, fai\_reopen, see below), shown next:

\begin{rcode}
> vcf_reopen( yvcfh )
(II) VCF version is 4.1
vcff::open : file opened, contains 526 samples
[1] TRUE
>
\end{rcode}

The VCF file is now usable again.
Before we can begin reading data from it, we need to take care of three points:

\begin{itemize}
	\item select the samples we are interested in (REQUIRED)
	\item set which region we are interested in (OPTIONAL)
	\item set up filtering step (OPTIONAL)
\end{itemize}

\subsection{Samples}

In order to know which samples we can select, we query their names:

\begin{rcode}
> GET SAMPLE NAMES
\end{rcode}


\begin{rcode}
> SELECT SAMPLES
\end{rcode}


\begin{rcode}
> GET SELECTED SAMPLES
\end{rcode}

Now that we have taken care of that, we can specify a region to get data from.
Regions are specified by three components: chromosome, start position, end position.
A VCF file can contain any number of chromosomes or contigs, provided that they have
unique identifiers.
\textcolor{red}{NOTE: the naming scheme of chromosomes can vary: sometimes they are named Chr1, Chr2,
ChrX, ChrY and sometimes it is just 1,2,X,Y. WhopGenome expects the correct identifier
and will not try to guess.}


%
%	regions, samples
%

%
%	filtering
%

\end{document}
